### PR TITLE
[DXIL] Add sign intrinsic part 2

### PR DESCRIPTION
**Pull request by tgymnich**
makes progress on #70078

### Changes
- Added `int_dx_sign` intrinsic in `IntrinsicsDirectX.td`
- Added expansion for `int_dx_sign in `DXILIntrinsicExpansion.cpp`
- Added DXIL backend test case

### Related PRs
- https://github.com/llvm/llvm-project/pull/101987
- https://github.com/llvm/llvm-project/pull/101989

### CODE REVIEW

**Review by farzonl — APPROVED** [review]

File: llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp
Comment (tgymnich) [reply llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:305]: We can either add `usign` builtins/intrinsics all the way down from hlsl to dxil and spirv. Or just add a high level builtin to hlsl. 

I don't like the idea of having an intrinsic for `select(x == 0, 0, 1)`.
I wonder from a consistency point of view if it is even necessary to have a `usign` builtin since `abs` also only seems to offer the signed variants in `hlsl_intrinsics.h`.

Comment (tgymnich) [reply llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp]: part 1 fixed in 61bc2a8

Comment (farzonl) [reply llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp]: i'm not opposed to doing a high level operation for unsigned  `sign` cases, however  for reference this is what DXC does: https://github.com/microsoft/DirectXShaderCompiler/blob/0e7591a6ee94c8c8eb0d536ce7815fd56a776451/lib/HLSL/HLOperationLower.cpp#L2703C1-L2714C2

**Comment by farzonl** [comment]
@tgymnich you got approval on this PR, is there something you are waiting on?

**Comment by tgymnich** [comment]
@farzonl This seems good to go. I cannot merge this myself though.

**Comment by tgymnich** [comment]
@farzonl PR should be good to go now. I don't have commit access, so I'd appreciate if you could merge this for me.

File: llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp
Comment (tgymnich) [reply llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:371]: Let's do the high level op then.

**Comment by farzonl** [comment]
> @farzonl This seems good to go. I cannot merge this myself though. I will take some time later today to rebase this PR on main.

Let me get one more reviewer from my team to take a look and then I'll merge it for you. Thanks again. Also thank you for your patience.

File: llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp
Comment (pow2clk) [reply llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:371]: > A small nitpick I like to keep the the builder setup code close together ie
> 
> ```c++
>  IRBuilder<> Builder(Orig->getParent());
> Builder.SetInsertPoint(Orig);
> ```
I didn't see any discussion about this nor any changes in code. Also peculiar to me is the resistance to use the more direct approach which is to set the insertion point with the constructor. Instead of setting it to the start of the basic block and then resetting it, you can just set it with `Orig` and be done with it. 

```c++
  IRBuilder<> Builder(Orig);
```

File: llvm/test/CodeGen/DirectX/sign.ll
Comment (pow2clk) [inline llvm/test/CodeGen/DirectX/sign.ll:214]: I don't follow. Is this file autogenerated? The checks below don't seem to be useful.

**Review by pow2clk — COMMENTED** [review]
I won't insist about the IRBuilder constructor issue, but the end of the test checks confuse me

File: llvm/test/CodeGen/DirectX/sign.ll
Comment (tgymnich) [reply llvm/test/CodeGen/DirectX/sign.ll]: removed. I might have run the test updater over this file at some point in the past.

**Comment by tgymnich** [comment]
@pow2clk I removed the unnecessary checks and fixed the `IRBuilder` usage.

**Comment by tgymnich** [comment]
I don't have commit access so feel free to merge this for me.

**Comment by farzonl** [comment]
> I don't have commit access so feel free to merge this for me.

I will. Our current policy on HLSL is two approvals. So I’m waiting for a second one.

**Review by llvm-beanz — APPROVED** [review]

**Review by pow2clk — APPROVED** [review]

**Comment by farzonl** [comment]
Thank you @tgymnich. Congrats on your first DirectX contribution!

**Pull request merged**
